### ISSUE TITLE
Continue improvements in the use date dialogue

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/R-Instat#4603]
1. The addition of the quarter function has been a good improvement.
2. The main suggestions I make now are to improve the layout of the check-boxes.
3. In addition, I think the Pentad does not work correctly.  It should produce a factor with 72 levels, and with just 5 (or 6) days in each month.  Perhaps Beth can look quickly at this?
4. And more of the variables should have an option of labels.  For example Quarter is just 1 2, 3, 4 which is correct if numeric.  It could be Jan-Mar, Apr-Jun, Jul-Sep, Oct-Dec as short labels.  And these should (like the months), ideally be done so that they are automatically in the computer's language where it is not English.  
5. Similarly Dekad would have the label option of Jan1, Jan2, Jan3, Feb1... With Pentad it would be Jan1 to Jan6. 
6. And more variables should be shifted as an option, namely Month, Pentad and Dekad.  When these are labelled they will look the same, but Aug would be level 1, if shifted to Aug, with July as level 12.  For Dekads Aug1 would become Level 1, with Jul1 being level 34.

So my suggested layout is roughly as shown by Danny who may take a  photo.  The 6 variables that can be shifted are first, each with a column, namely 
**Year, Month, Day in Year(366), Pentad, Dekad, Quarter.**
In the left hand side are the 3 options **Numeric, Abb Name, Name**, with check-boxes to indicate what is possible.
So:
a) Year is just numeric,
b) Month is Numeric, or Abb Name or Name.
c) Day in Year(366) is just numeric
d) Pentad, Dekad, Quarter are Numeric or Abb Name.

Below that block is:
**Shifted Month and pull-down** as we have.

Below that is another block of variables to which shifted does not apply, i.e. 
a) Day in Month - just numeric
b) Day in Year - just numeric
c) Week is numeric or abbrev name Wk1, Wk52 or full name Week1, to Week52.
d) Weekday also has the 3 options
e) Leap Year is just numeric (it is actually logical, but let's not be too purist!

I suggest that Patrick could work on the layout and organisation of the dialogue, while Beth concentrates (with him) on the R code.

**Comment by bethanclarke** [comment]
I’ve been working on use date dialog in R, just looking at the code so far. 

I’ve created a month column which is a factor where 1 is Aug, 2 Sept etc. to July 12- I’ve just don’t this by relabelling the levels. I created another month factor column produced using fct_shift() command from the forcats package which has the following levels: Aug > Sept ... >July. 

For pentads I found another function to create a variable that has 5 days in a week, 6 weeks in a month and 72 weeks in the year. The function in the data object wasn’t producing the correct vector. Dekads works just fine.

**Comment by Patowhiz** [comment]
@rdstern @dannyparsons what will we be shifting in Day in Year if the day control is to be removed from the form? And what will be the group header names of the 2 sections?
Below is a sample design of what's proposed. @bethanclarke could you have the R code merged if its ready?

![new use date design](https://user-images.githubusercontent.com/21179007/40354010-8d461fe6-5dbb-11e8-99ac-d91b04a44a80.JPG)

**Comment by rdstern** [comment]
Patrick, my first impression is that this looks very nice.  
Check with Danny on the Day control.  With this layout I would now be happy to keep it as you have it. Default is 1.
But it only makes sense for the (Shifted) Year and the (Shifted) Day of Year.  So it is hidden (Iand not part of the R code) for Month, Dekad, Pentad or Quarter.

**Comment by bethanclarke** [comment]
@patowhiz would you like me to send you the R Code or for me to implement it in the dialog?

@dannyparsons, was the R-Code I sent what we were looking for?

**Comment by rdstern** [comment]
Beth, I suggest you implementing might be more efficient time-wise.  Then Patrick can proceed with other tasks and then come back to this dialogue later to test and then help his understanding.

**Comment by rdstern** [comment]
Currently I like the blocks, partly because they emphasise the possibility of shifting for the first set and not for the others.  I don't have a good name to distinguish the sets, so let's not bother for now.

**Comment by bethanclarke** [comment]
@patowhiz I can include the new R code for pentads in the data object but I’ll have to check will Danny about the other lines of code

**Comment by Patowhiz** [comment]
@bethanclarke okay. I would be finalising the design discussion with danny once he arrives then I could share the design code changes with you for the implementation as @rdstern has suggested above.

**Comment by Patowhiz** [comment]
Final design of the dialog.
![use date design](https://user-images.githubusercontent.com/21179007/40499851-d7c62da0-5f8b-11e8-9229-cfa8a1c6ae26.JPG)

**Comment by Patowhiz** [comment]
Final design of the dialog.

![usedate](https://user-images.githubusercontent.com/21179007/40500002-3cf59b5c-5f8c-11e8-8cd9-9f5335872f04.JPG)

**Comment by rdstern** [comment]
Looks good.  What are those funny symbols (or half-symbols) for example after the checkbox in Day in Year, Numeric?

**Comment by Patowhiz** [comment]
It's just my screen resolution doing that. Though they will disappear
during runtime.

**Comment by rdstern** [comment]
To complete the dialogue needs Beth's code.  Could Patrick (for now) disable the Pentade and the Dekad (with label) options, and then get the rest of the dialogue working - presumably using the old code.  Then Beth can share the R code for the other options, or add those options herself later?

**Comment by Patowhiz** [comment]
@rdstern @dannyparsons @bethanclarke  this is now ready. However a few points to note.
Shifting of pentad,dekad does not work - Rcode missing.
Pentad,dekad,quarter abbreviations and  also week abbreviation and name are disabled - will be enabled once Rcode is ready.
The image below illustrates the controls disabled.

I have also modified the R code to take up the new parameters and to adapt to the new changes. So it will just be a matter of enabling the controls once the pending R code additions are added.

@rdstern the resulting column names for numerics well have val appended to them e.g week_val except for pentad,dekad and quarter(for now, since they don't have abbr functionalities yet).

![dlgusedate](https://user-images.githubusercontent.com/21179007/40773857-c52c96ca-64cc-11e8-929e-a2a9b135a4d1.JPG)

**Comment by dannyparsons** [comment]
This has made good progress. I suggest @bethanclarke now takes over the pull request to add the remaining functions before it is merged since it's not fully functional yet.

**Comment by rdstern** [comment]
There is one additional item that would be useful here and that Patrick might do, before the "official handover" to Beth.  That is to dictate the order that the resulting variables are produced.  To me there is an obvious order and that is not too far from the current order of the controls. 

While there could we move Quarter so it is before Dekad in the "top group"?

So I suggest either: 

a) That they go in that order, i.e. first 6 and then the next 5.
b) Year, then Month then Day in Month (from 2nd set), then Day in Year (366) then Day in Year.  Then in order from there, i.e. Quarter, Dekad, Pentad, Week, DayinWeek, Leap Year.  

Once this is done, then, with the missing ones (pentad, dekad with labels) currently disabled, then could Beth both test and then also add the missing options.

But it would be good to merge the resulting dialogue soon.  So, if the new calculations will take some time, then could the new layout be merged with some of the options still disabled?

**Comment by bethanclarke** [comment]
@Patowhiz have you made the changes in Roger's last message?

**Comment by Patowhiz** [comment]
@rdstern I'm not sure about the proposed structure in your item (b) perhaps @dannyparsons  can comment on it. (What about the aspect of grouping based on variables that can be shifted and those that can't, that was the original idea).
@bethanclarke no, I haven't changed the position of the quarter in the design i.e according to Roger's new design proposal.

**Comment by rdstern** [comment]
My item b) has no change on the dialogue.  It is concerned only with the order of resulting columns that are produced and added to the data frame.  I suggested a logical order to them.  Currently when you opt for both month_abbr and year for example, the month is produced first and the year is a variable after the month.

**Comment by Patowhiz** [comment]
Okay. That's fine. I can have that done quickly

**Comment by Patowhiz** [comment]
@rdstern I've set the resulting columns to be arranged in the following order; leap year, year ,month, day in month, day in year 366, day in year, quarter, dekad, pentad, week, weekday. (I wasn't quite sure if you wanted the year to be the last or to be grouped with the other year value).

@bethanclarke the changes have now been made.
Below is a screenshot of the new design.

![use date](https://user-images.githubusercontent.com/21179007/41001661-4511b234-691a-11e8-9848-8dcb567219b6.JPG)

**Comment by Patowhiz** [comment]
@rdstern  this is more of a question, How is shifting of the month column supposed  to work, I mean what will be the expected result? The shifting of the month is currently not implemented in the R code.

**Comment by rdstern** [comment]
It will look the same.  But the Month is an ordered factor, (and also circular) with (by default) Jan having level 1, Feb is 2 and Dec is 12.  If shifted to August, then August will be level 1 and July will be level 12.  Then tables and graphs will go from August to July.  It is the function fct_shift in the forcats package.

**Comment by Patowhiz** [comment]
Now i get it. Thank you

**Comment by rdstern** [comment]
The workshop has introduced another issue in this dialogue, and that concerns the Year column, when the date is shifted.
Currently the Year column is numeric, e.g. 1945, if the year is not shifted.  If shifted then it is a factor, e.g. 1944-1945.  This Malawi workshop was the first one that used the shifted year (and day of year) as routine.  That's because the rainy season is from October to April.

I first thought that the year as a factor was going to be problematic with the line graphs.  I still think it will be inconvenient for some of the graphs, but was impressed how well it worked.  One inconvenience for me is that the x-axis is labelled with each year - that's a problem we had with boxplots earlier.  However, in Malawi, they have been getting their graphs using Excel so far and they all has this same feature.  So they were quite happy.

Back to this dialogue.  I suggest one solution is to have the shifted year as numeric (giving perhaps the first year, so 1944-1945 becomes 1944.  (It is the year in which the period starts.  Then the current 1944-1945 could become the full name.  I wonder if then either 1944-45, then 1999-00, becomes the abbreviated label, or even 44-45.

**Comment by dannyparsons** [comment]
That sounds right, that's what I assumed it was already doing for shifted year.

**Comment by Patowhiz** [comment]
I'm not getting your question. I used checkboxes and the table control

**Comment by Patowhiz** [comment]
I think what's remaining here is the Quarter abbreviation and week abbreviation and week name. They are the ones disabled after the last PR that addressed this issue.

**Comment by rdstern** [comment]
1. A small improvement here is to the dekads and pentades.  They now generate correctly, but the abbreviated versions are character, rather than factor.  Please could they be factor.  (It is currently difficult to make them a factor, because they come in alphabetical order, when they should be in the period in the year-order.  (And fixing that is a pain, because there are so many levels.)
As with the months this should also be possible for the shifted data.  They generate correctly, so it is just a question of making the factor. 
2.  The months are currently correct.  But the factor for the months is currently made into an ordered factor.  This is true, but usually not what is wanted.  So please could that change to be an ordinary factor.
